// mongodb-mcp-server/dist/cjs/tools/mongodb/create/createIndex.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateIndexTool = void 0;
const zod_1 = require("zod");
const mongodbTool_js_1 = require("../mongodbTool.js");
class CreateIndexTool extends mongodbTool_js_1.MongoDBToolBase {
    constructor() {
        super(...arguments);
        this.name = "create-index";
        this.description = "Create an index for a collection";
        this.argsShape = {
            ...mongodbTool_js_1.DbOperationArgs,
            keys: zod_1.z.object({}).catchall(zod_1.z.custom()).describe("The index definition"),
            name: zod_1.z.string().optional().describe("The name of the index"),
        };
        this.operationType = "create";
    }
    async execute({ database, collection, keys, name, }) {
        const provider = await this.ensureConnected();
        const indexes = await provider.createIndexes(database, collection, [
            {
                key: keys,
                name,
            },
        ]);
        return {
            content: [
                {
                    text: `Created the index "${indexes[0]}" on collection "${collection}" in database "${database}"`,
                    type: "text",
                },
            ],
        };
    }
}
exports.CreateIndexTool = CreateIndexTool;
