// mongodb-mcp-server/dist/cjs/tools/mongodb/create/createCollection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateCollectionTool = void 0;
const mongodbTool_js_1 = require("../mongodbTool.js");
class CreateCollectionTool extends mongodbTool_js_1.MongoDBToolBase {
    constructor() {
        super(...arguments);
        this.name = "create-collection";
        this.description = "Creates a new collection in a database. If the database doesn't exist, it will be created automatically.";
        this.argsShape = mongodbTool_js_1.DbOperationArgs;
        this.operationType = "create";
    }
    async execute({ collection, database }) {
        const provider = await this.ensureConnected();
        await provider.createCollection(database, collection);
        return {
            content: [
                {
                    type: "text",
                    text: `Collection "${collection}" created in database "${database}".`,
                },
            ],
        };
    }
}
exports.CreateCollectionTool = CreateCollectionTool;
