// mongodb-mcp-server/dist/cjs/tools/mongodb/metadata/dbStats.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DbStatsTool = void 0;
const mongodbTool_js_1 = require("../mongodbTool.js");
const tool_js_1 = require("../../tool.js");
const bson_1 = require("bson");
class DbStatsTool extends mongodbTool_js_1.MongoDBToolBase {
    constructor() {
        super(...arguments);
        this.name = "db-stats";
        this.description = "Returns statistics that reflect the use state of a single database";
        this.argsShape = {
            database: mongodbTool_js_1.DbOperationArgs.database,
        };
        this.operationType = "metadata";
    }
    async execute({ database }) {
        const provider = await this.ensureConnected();
        const result = await provider.runCommandWithCheck(database, {
            dbStats: 1,
            scale: 1,
        });
        return {
            content: (0, tool_js_1.formatUntrustedData)(`Statistics for database ${database}`, bson_1.EJSON.stringify(result)),
        };
    }
}
exports.DbStatsTool = DbStatsTool;
