// mongodb-mcp-server/dist/esm/tools/atlas/read/listOrgs.js
import { AtlasToolBase } from "../atlasTool.js";
import { formatUntrustedData } from "../../tool.js";
export class ListOrganizationsTool extends AtlasToolBase {
    constructor() {
        super(...arguments);
        this.name = "atlas-list-orgs";
        this.description = "List MongoDB Atlas organizations";
        this.operationType = "read";
        this.argsShape = {};
    }
    async execute() {
        const data = await this.session.apiClient.listOrganizations();
        if (!data?.results?.length) {
            return {
                content: [{ type: "text", text: "No organizations found in your MongoDB Atlas account." }],
            };
        }
        // Format organizations as a table
        const output = `Organization Name | Organization ID
----------------| ----------------
` +
            data.results
                .map((org) => {
                return `${org.name} | ${org.id}`;
            })
                .join("\n");
        return {
            content: formatUntrustedData(`Found ${data.results.length} organizations in your MongoDB Atlas account.`, output),
        };
    }
}
