// mongodb-mcp-server/dist/esm/tools/mongodb/read/collectionIndexes.js
import { DbOperationArgs, MongoDBToolBase } from "../mongodbTool.js";
import { formatUntrustedData } from "../../tool.js";
export class CollectionIndexesTool extends MongoDBToolBase {
    constructor() {
        super(...arguments);
        this.name = "collection-indexes";
        this.description = "Describe the indexes for a collection";
        this.argsShape = DbOperationArgs;
        this.operationType = "read";
    }
    async execute({ database, collection }) {
        const provider = await this.ensureConnected();
        const indexes = await provider.getIndexes(database, collection);
        return {
            content: formatUntrustedData(`Found ${indexes.length} indexes in the collection "${collection}":`, indexes.length > 0
                ? indexes
                    .map((index) => `Name: "${index.name}", definition: ${JSON.stringify(index.key)}`)
                    .join("\n")
                : undefined),
        };
    }
    handleError(error, args) {
        if (error instanceof Error && "codeName" in error && error.codeName === "NamespaceNotFound") {
            return {
                content: [
                    {
                        text: `The indexes for "${args.database}.${args.collection}" cannot be determined because the collection does not exist.`,
                        type: "text",
                    },
                ],
            };
        }
        return super.handleError(error, args);
    }
}
