// mongodb-mcp-server/dist/cjs/tools/mongodb/delete/deleteMany.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeleteManyTool = void 0;
const zod_1 = require("zod");
const mongodbTool_js_1 = require("../mongodbTool.js");
const indexCheck_js_1 = require("../../../helpers/indexCheck.js");
class DeleteManyTool extends mongodbTool_js_1.MongoDBToolBase {
    constructor() {
        super(...arguments);
        this.name = "delete-many";
        this.description = "Removes all documents that match the filter from a MongoDB collection";
        this.argsShape = {
            ...mongodbTool_js_1.DbOperationArgs,
            filter: zod_1.z
                .object({})
                .passthrough()
                .optional()
                .describe("The query filter, specifying the deletion criteria. Matches the syntax of the filter argument of db.collection.deleteMany()"),
        };
        this.operationType = "delete";
    }
    async execute({ database, collection, filter, }) {
        const provider = await this.ensureConnected();
        // Check if delete operation uses an index if enabled
        if (this.config.indexCheck) {
            await (0, indexCheck_js_1.checkIndexUsage)(provider, database, collection, "deleteMany", async () => {
                return provider.runCommandWithCheck(database, {
                    explain: {
                        delete: collection,
                        deletes: [
                            {
                                q: filter || {},
                                limit: 0, // 0 means delete all matching documents
                            },
                        ],
                    },
                    verbosity: "queryPlanner",
                });
            });
        }
        const result = await provider.deleteMany(database, collection, filter);
        return {
            content: [
                {
                    text: `Deleted \`${result.deletedCount}\` document(s) from collection "${collection}"`,
                    type: "text",
                },
            ],
        };
    }
}
exports.DeleteManyTool = DeleteManyTool;
