// mongodb-mcp-server/dist/cjs/tools/mongodb/delete/dropDatabase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DropDatabaseTool = void 0;
const mongodbTool_js_1 = require("../mongodbTool.js");
class DropDatabaseTool extends mongodbTool_js_1.MongoDBToolBase {
    constructor() {
        super(...arguments);
        this.name = "drop-database";
        this.description = "Removes the specified database, deleting the associated data files";
        this.argsShape = {
            database: mongodbTool_js_1.DbOperationArgs.database,
        };
        this.operationType = "delete";
    }
    async execute({ database }) {
        const provider = await this.ensureConnected();
        const result = await provider.dropDatabase(database);
        return {
            content: [
                {
                    text: `${result.ok ? "Successfully dropped" : "Failed to drop"} database "${database}"`,
                    type: "text",
                },
            ],
        };
    }
}
exports.DropDatabaseTool = DropDatabaseTool;
