// mongodb-mcp-server/dist/cjs/tools/mongodb/delete/dropCollection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DropCollectionTool = void 0;
const mongodbTool_js_1 = require("../mongodbTool.js");
class DropCollectionTool extends mongodbTool_js_1.MongoDBToolBase {
    constructor() {
        super(...arguments);
        this.name = "drop-collection";
        this.description = "Removes a collection or view from the database. The method also removes any indexes associated with the dropped collection.";
        this.argsShape = {
            ...mongodbTool_js_1.DbOperationArgs,
        };
        this.operationType = "delete";
    }
    async execute({ database, collection }) {
        const provider = await this.ensureConnected();
        const result = await provider.dropCollection(database, collection);
        return {
            content: [
                {
                    text: `${result ? "Successfully dropped" : "Failed to drop"} collection "${collection}" from database "${database}"`,
                    type: "text",
                },
            ],
        };
    }
}
exports.DropCollectionTool = DropCollectionTool;
