// mongodb-mcp-server/dist/cjs/tools/atlas/create/createFreeCluster.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateFreeClusterTool = void 0;
const zod_1 = require("zod");
const atlasTool_js_1 = require("../atlasTool.js");
const accessListUtils_js_1 = require("../../../common/atlas/accessListUtils.js");
class CreateFreeClusterTool extends atlasTool_js_1.AtlasToolBase {
    constructor() {
        super(...arguments);
        this.name = "atlas-create-free-cluster";
        this.description = "Create a free MongoDB Atlas cluster";
        this.operationType = "create";
        this.argsShape = {
            projectId: zod_1.z.string().describe("Atlas project ID to create the cluster in"),
            name: zod_1.z.string().describe("Name of the cluster"),
            region: zod_1.z.string().describe("Region of the cluster").default("US_EAST_1"),
        };
    }
    async execute({ projectId, name, region }) {
        const input = {
            groupId: projectId,
            name,
            clusterType: "REPLICASET",
            replicationSpecs: [
                {
                    zoneName: "Zone 1",
                    regionConfigs: [
                        {
                            providerName: "TENANT",
                            backingProviderName: "AWS",
                            regionName: region,
                            electableSpecs: {
                                instanceSize: "M0",
                            },
                        },
                    ],
                },
            ],
            terminationProtectionEnabled: false,
        };
        await (0, accessListUtils_js_1.ensureCurrentIpInAccessList)(this.session.apiClient, projectId);
        await this.session.apiClient.createCluster({
            params: {
                path: {
                    groupId: projectId,
                },
            },
            body: input,
        });
        return {
            content: [
                { type: "text", text: `Cluster "${name}" has been created in region "${region}".` },
                { type: "text", text: `Double check your access lists to enable your current IP.` },
            ],
        };
    }
}
exports.CreateFreeClusterTool = CreateFreeClusterTool;
