// mongodb-mcp-server/dist/cjs/tools/mongodb/update/renameCollection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RenameCollectionTool = void 0;
const zod_1 = require("zod");
const mongodbTool_js_1 = require("../mongodbTool.js");
class RenameCollectionTool extends mongodbTool_js_1.MongoDBToolBase {
    constructor() {
        super(...arguments);
        this.name = "rename-collection";
        this.description = "Renames a collection in a MongoDB database";
        this.argsShape = {
            ...mongodbTool_js_1.DbOperationArgs,
            newName: zod_1.z.string().describe("The new name for the collection"),
            dropTarget: zod_1.z.boolean().optional().default(false).describe("If true, drops the target collection if it exists"),
        };
        this.operationType = "update";
    }
    async execute({ database, collection, newName, dropTarget, }) {
        const provider = await this.ensureConnected();
        const result = await provider.renameCollection(database, collection, newName, {
            dropTarget,
        });
        return {
            content: [
                {
                    text: `Collection "${collection}" renamed to "${result.collectionName}" in database "${database}".`,
                    type: "text",
                },
            ],
        };
    }
    handleError(error, args) {
        if (error instanceof Error && "codeName" in error) {
            switch (error.codeName) {
                case "NamespaceNotFound":
                    return {
                        content: [
                            {
                                text: `Cannot rename "${args.database}.${args.collection}" because it doesn't exist.`,
                                type: "text",
                            },
                        ],
                    };
                case "NamespaceExists":
                    return {
                        content: [
                            {
                                text: `Cannot rename "${args.database}.${args.collection}" to "${args.newName}" because the target collection already exists. If you want to overwrite it, set the "dropTarget" argument to true.`,
                                type: "text",
                            },
                        ],
                    };
            }
        }
        return super.handleError(error, args);
    }
}
exports.RenameCollectionTool = RenameCollectionTool;
