// mongodb-mcp-server/dist/cjs/tools/atlas/read/inspectAccessList.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InspectAccessListTool = void 0;
const zod_1 = require("zod");
const atlasTool_js_1 = require("../atlasTool.js");
const tool_js_1 = require("../../tool.js");
class InspectAccessListTool extends atlasTool_js_1.AtlasToolBase {
    constructor() {
        super(...arguments);
        this.name = "atlas-inspect-access-list";
        this.description = "Inspect Ip/CIDR ranges with access to your MongoDB Atlas clusters.";
        this.operationType = "read";
        this.argsShape = {
            projectId: zod_1.z.string().describe("Atlas project ID"),
        };
    }
    async execute({ projectId }) {
        const accessList = await this.session.apiClient.listProjectIpAccessLists({
            params: {
                path: {
                    groupId: projectId,
                },
            },
        });
        const results = accessList.results ?? [];
        if (!results.length) {
            return {
                content: [{ type: "text", text: "No access list entries found." }],
            };
        }
        return {
            content: (0, tool_js_1.formatUntrustedData)(`Found ${results.length} access list entries`, `IP ADDRESS | CIDR | COMMENT
------|------|------
${results
                .map((entry) => {
                return `${entry.ipAddress} | ${entry.cidrBlock} | ${entry.comment}`;
            })
                .join("\n")}`),
        };
    }
}
exports.InspectAccessListTool = InspectAccessListTool;
