// mongodb-mcp-server/dist/cjs/tools/atlas/read/listProjects.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListProjectsTool = void 0;
const atlasTool_js_1 = require("../atlasTool.js");
const tool_js_1 = require("../../tool.js");
const zod_1 = require("zod");
class ListProjectsTool extends atlasTool_js_1.AtlasToolBase {
    constructor() {
        super(...arguments);
        this.name = "atlas-list-projects";
        this.description = "List MongoDB Atlas projects";
        this.operationType = "read";
        this.argsShape = {
            orgId: zod_1.z.string().describe("Atlas organization ID to filter projects").optional(),
        };
    }
    async execute({ orgId }) {
        const orgData = await this.session.apiClient.listOrganizations();
        if (!orgData?.results?.length) {
            return {
                content: [{ type: "text", text: "No organizations found in your MongoDB Atlas account." }],
            };
        }
        const orgs = orgData.results
            .map((org) => [org.id || "", org.name])
            .filter(([id]) => id)
            .reduce((acc, [id, name]) => ({ ...acc, [id]: name }), {});
        const data = orgId
            ? await this.session.apiClient.listOrganizationProjects({
                params: {
                    path: {
                        orgId,
                    },
                },
            })
            : await this.session.apiClient.listProjects();
        if (!data?.results?.length) {
            return {
                content: [{ type: "text", text: `No projects found in organization ${orgId}.` }],
            };
        }
        // Format projects as a table
        const rows = data.results
            .map((project) => {
            const createdAt = project.created ? new Date(project.created).toLocaleString() : "N/A";
            const orgName = orgs[project.orgId] ?? "N/A";
            return `${project.name} | ${project.id} | ${orgName} | ${project.orgId} | ${createdAt}`;
        })
            .join("\n");
        const formattedProjects = `Project Name | Project ID | Organization Name | Organization ID | Created At
----------------| ----------------| ----------------| ----------------| ----------------
${rows}`;
        return {
            content: (0, tool_js_1.formatUntrustedData)(`Found ${rows.length} projects`, formattedProjects),
        };
    }
}
exports.ListProjectsTool = ListProjectsTool;
