// mongodb-mcp-server/dist/cjs/tools/atlas/read/listOrgs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListOrganizationsTool = void 0;
const atlasTool_js_1 = require("../atlasTool.js");
const tool_js_1 = require("../../tool.js");
class ListOrganizationsTool extends atlasTool_js_1.AtlasToolBase {
    constructor() {
        super(...arguments);
        this.name = "atlas-list-orgs";
        this.description = "List MongoDB Atlas organizations";
        this.operationType = "read";
        this.argsShape = {};
    }
    async execute() {
        const data = await this.session.apiClient.listOrganizations();
        if (!data?.results?.length) {
            return {
                content: [{ type: "text", text: "No organizations found in your MongoDB Atlas account." }],
            };
        }
        // Format organizations as a table
        const output = `Organization Name | Organization ID
----------------| ----------------
` +
            data.results
                .map((org) => {
                return `${org.name} | ${org.id}`;
            })
                .join("\n");
        return {
            content: (0, tool_js_1.formatUntrustedData)(`Found ${data.results.length} organizations in your MongoDB Atlas account.`, output),
        };
    }
}
exports.ListOrganizationsTool = ListOrganizationsTool;
